devtools/zenodo/zenodo_data_release.py

```diff
@@ -0,0 +1,295 @@
+            }
+        elif env == PRODUCTION:
+            self.base_url = "https://zenodo.org/api"
+            self.auth_headers = {
+                "Authorization": f"Bearer {os.environ['ZENODO_TOKEN_PUBLISH']}"
+            }
+        else:
+            raise ValueError(
+                f"Got unexpected {env=}, expected {SANDBOX} or {PRODUCTION}"
+            )
+
+    def get_deposition(self, deposition_id: int) -> requests.Response:
+        """LEGACY API: Get JSON describing a deposition.
+
+        Depositions can be published *or* unpublished.
+        """
+        return requests.get(
+            f"{self.base_url}/deposit/depositions/{deposition_id}",
+            headers=self.auth_headers,
+            timeout=10,
+        )
+
+    def get_record(self, record_id: int) -> requests.Response:
+        """NEW API: Get JSON describing a record.
+
+        All records are published records.
+        """
+        return requests.get(
+            f"{self.base_url}/records/{record_id}",
+            headers=self.auth_headers,
+            timeout=10,
+        )
+
+    def new_record_version(self, record_id: int) -> requests.Response:
+        """NEW API: get or create the draft associated with a record ID.
+
+        Finds the latest record in the concept that record_id points to, and
+        makes a new version unless one exists already.
+        """
+        return requests.post(
+            f"{self.base_url}/records/{record_id}/versions",
+            headers=self.auth_headers,
+            timeout=10,
+        )
+
+    def update_deposition_metadata(
+        self, deposition_id: int, metadata: dict
+    ) -> requests.Response:
+        """LEGACY API: Update deposition metadata.
+
+        Replaces the existing metadata completely - so make sure to pass in
+        complete metadata. You cannot update metadata fields one at a time.
+        """
+        url = f"{self.base_url}/deposit/depositions/{deposition_id}"
+        data = {"metadata": metadata}
+        response = requests.put(url, json=data, headers=self.auth_headers, timeout=10)
+        return response
+
+    def delete_deposition_file(self, deposition_id: int, file_id) -> requests.Response:
+        """LEGACY API: Delete file from deposition.
+
+        Note: file_id is not always the file name.
+        """
+        return requests.delete(
+            f"{self.base_url}/deposit/depositions/{deposition_id}/files/{file_id}",
+            headers=self.auth_headers,
+            timeout=10,
+        )
+
+    def create_deposition_file(
+        self, deposition_id: int, file_name: str, file_content: BinaryIO
+    ) -> requests.Response:
+        """LEGACY API: Create file in deposition."""
+        url = f"{self.base_url}/deposit/depositions/{deposition_id}/files"
+        files = {"file": file_content, "name": file_name}
+        response = requests.post(
+            url,
+            headers=self.auth_headers,
+            files=files,
+            timeout=1_000,
```

Comment:
If it takes longer than 1000 seconds to upload one of the large files what happens?

If any of the files fail to upload on the first try, does the creation of the new archive necessarily fail? Or can individual files be retried if they don't work the first time?

Reply:
If it takes longer than 1,000 seconds to upload one of the large files, it will time out! We can set this to some ridiculously high number.

Currently there aren't retries.

We could add retry handling and also pre-publishing hash checks for sure, do you think that's blocking this merge or more of a follow-up sort of situation?

Reply:
I'm trying to think of the things that I would check as a human who had just run the script in reviewing the draft.

My expectation is that most of the time we will get some failures when uploading a new archive, usually on the larger files, and will have to retry.

We can hold off the retry logic and try running the script to see what happens, but I think we should clearly define what we want a failure to look like.

- Should it leave the files that did upload successfully in a draft? Or should it wipe everything and set it back to a no-draft state?
- Should it keep going and attempt to upload additional files after one of them has failed? Or should it bail as soon as there's a failure?
- If a failure is detected, it should never publish the archive, even if `--publish` has been set.

For now while we're testing and developing it, if one of the files fails to upload I kinda thing we still want it to iterate through the other files and try at least once to get them uploaded so we can see what happens with all of them.  And then leave the draft in whatever state it ended up in so we can inspect it.